### PR TITLE
feat(dialog, dialog-full-screen): add help and and fix text wrapping - [ FE-4025 | FE-4026 ] 

### LINKED ISSUE #3828: Dialog : Header style when too long
### Current behaviour
When a title is too long, it exceeds the size of the dialog

### Expected behaviour
Either the title is truncated (with '...') or displayed on multiple lines.

### Reproducible example
image

Reproducible in the storybook

### Your environment

| Software         | Version(s) |
| ---------------- | ---------- |
| carbon-react     | 68.0.2 |

### LINKED ISSUE #3831: Add prop to Dialog to enable a help link in the heading
### Current behaviour
The `help link` is placed on the next line

### Expected behaviour
The `help link` have to be aligned to the right side of the title (this dialog header is custom styled without a `Heading` component)

### Reproducible example
https://codesandbox.io/s/jovial-sammet-hp90r

### Additional context
This codesandbox uses carbon `42.6.0` but this problem also exists in version `62.0.0`

### Your environment

| Software         | Version(s) |
| ---------------- | ---------- |
| carbon-react     |  62.0.0
| carbon-factory   |
| react-scripts    | 
| React            | 16.13.1
| Browser          | Chrome / Firefox
| npm              |
| Operating System | MacOS

### PR DESCRIPTION
**Pull request by dawidzarzycki**
### Proposed behaviour
- add help prop 
- wrap text
dialog_dialog-full-screen_help

Fixes #3828
Fixes #3831

### Current behaviour
help prop doesn't exist

### Checklist

- [x] Commits follow our style guide
- [x] Screenshots are included in the PR if useful
- [ ] <del> All themes are supported if required
- [x] Unit tests added or updated if required
- [ ] Cypress automation tests added or updated if required
- [ ] Storybook added or updated if required
- [ ] Typescript `d.ts` file added or updated if required
- [ ] <del> Carbon implementation and Design System documentation are congruent

### Testing instructions

DialogFullScreen: https://5ecf782fe724630022d27d7d-oskymkvuxe.chromatic.com/?path=/story/dialog-full-screen--with-help
Dialog: https://5ecf782fe724630022d27d7d-oskymkvuxe.chromatic.com/?path=/story/dialog--with-help

### CODE REVIEW

File: src/components/dialog/dialog.stories.mdx [hunk 1]
```diff
@@ -168,6 +168,52 @@ When mixing editable and non-editable content, you can use the <LinkTo kind='Des
   </Story>
 </Preview>
 
+### With help 
+
+<Preview>
+  <Story name="help" parameters={{ chromatic: { disable: true } }}>
```
Comment (edleeks87) [inline src/components/dialog/dialog.stories.mdx:174]: `with help` would match the naming on the fullscreen example, I also notice this is disabled for chromatic but the fullscreen example isn't I think either both should either be snapshotted or disabled

**Review by edleeks87 — COMMENTED** [review]
Looks good just a **very** minor comment from me

File: src/components/dialog-full-screen/dialog-full-screen.stories.mdx [hunk 2]
```diff
@@ -660,6 +660,55 @@ to have a possibility to manage active `Tabs` group
...
 </Preview>
 
+### With help
+
+<Preview>
+  <Story name="with help" parameters={{ viewports: [500, 1400] }}>
+    {() => {
+      const [isOpen, setIsOpen] = useState(true);
+      const aboveBreakpoint = useMediaQuery("(min-width: 568px)");
+      const verticalMargin = aboveBreakpoint ? "26px" : 0;
```
Comment (sjohnson-sage) [inline src/components/dialog-full-screen/dialog-full-screen.stories.mdx:670]: I don't think we need this breakpoint stuff in this story

File: src/components/dialog/dialog.stories.mdx [hunk 3]
```diff
@@ -168,6 +168,52 @@ When mixing editable and non-editable content, you can use the <LinkTo kind='Des
...
+    {() => {
+      const [isOpen, setIsOpen] = useState(false);
+      const [isDisabled, setIsDisabled] = useState(true);
```
Comment (sjohnson-sage) [inline src/components/dialog/dialog.stories.mdx:177]: `isDisabled` and `radioValue` can be removed here

File: src/components/dialog-full-screen/dialog-full-screen.stories.mdx [hunk 4]
```diff
@@ -660,6 +660,55 @@ to have a possibility to manage active `Tabs` group
...
+        <>
+          <Button onClick={() => setIsOpen(!isOpen)}>
+            Open DialogFullScreen
+          </Button>
+          <DialogFullScreen
+            open={isOpen}
+            onCancel={() => setIsOpen(false)}
+            title="An example of a long header"
+            subtitle="Subtitle"
+            help="Test help text"
```
Comment (sjohnson-sage) [inline src/components/dialog-full-screen/dialog-full-screen.stories.mdx:681]:
```suggestion
            help="Some help text"
```

or something similar that doesn't involve "test"

File: src/components/dialog/dialog.stories.mdx [hunk 5]
```diff
@@ -168,6 +168,52 @@ When mixing editable and non-editable content, you can use the <LinkTo kind='Des
...
+      const [radioValue, setRadioValue] = useState("1");
+      return (
+        <>
+          <Button onClick={() => setIsOpen(!isOpen)}>Open Dialog</Button>
+          <Dialog
+            open={isOpen}
+            onCancel={() => setIsOpen(false)}
+            title="Add an address"
+            help="Test help text"
```
Comment (sjohnson-sage) [inline src/components/dialog/dialog.stories.mdx:186]: Same as above

File: src/components/dialog-full-screen/dialog-full-screen.stories.mdx [hunk 6]
```diff
@@ -660,6 +660,55 @@ to have a possibility to manage active `Tabs` group
   </Story>
 </Preview>
 
+### With help
+
+<Preview>
+  <Story name="with help" parameters={{ viewports: [500, 1400] }}>
```
Comment (sjohnson-sage) [inline src/components/dialog-full-screen/dialog-full-screen.stories.mdx:666]: I don't think we need the viewports here, that was just for responsive stuff IIRC

File: src/components/dialog/dialog.style.js [hunk 7]
```diff
@@ -84,14 +88,17 @@ const DialogTitleStyle = styled.div`
...
+  }
+
   ${StyledHeading} {
     margin-bottom: 20px;
 
     ${StyledHeadingTitle} {
       color: ${({ theme }) => theme.text.color};
       display: block;
       overflow: hidden;
-      white-space: nowrap;
```
Comment (sjohnson-sage) [inline src/components/dialog/dialog.style.js:94]: How come this needed to be removed?

**Review by sjohnson-sage — COMMENTED** [review]
Looks good, just a couple of minor comments!

File: src/components/dialog/dialog.style.js (hunk 7 above)
Comment (dawidzarzycki) [reply src/components/dialog/dialog.style.js:94]: It helps with text wrapping

Comment (sjohnson-sage) [reply src/components/dialog/dialog.style.js:94]: Sorry I didn't read the part of the PR about fixing text wrapping!

**Review by sjohnson-sage — APPROVED** [review]

**Review by edleeks87 — APPROVED** [review]

**Review by edleeks87 — DISMISSED** [review]

**Review by sjohnson-sage — APPROVED** [review]

**Review by edleeks87 — DISMISSED** [review]

**Review by edleeks87 — DISMISSED** [review]

**Review by sjohnson-sage — DISMISSED** [review]

**Comment by harpalsingh** [comment]
all good, but I think we will take a look from a design perspective later about the placement of the help icon when text wraps, but I will work with the DS team on this, as it current stays on the right hand side, even after the text wraps onto the 2nd line. (FYI it might be problematic to add to the end of the text unless its built into the H tag.)

**Review by edleeks87 — DISMISSED** [review]

File: src/components/dialog/dialog.component.js [hunk 8]
```diff
@@ -177,6 +179,7 @@ const Dialog = ({
         focusFirstElement={focusFirstElement}
         bespokeTrap={bespokeFocusTrap}
         wrapperRef={dialogRef}
+        isHelp={help}
```
Comment (edleeks87) [inline src/components/dialog/dialog.component.js:182]: Is this needed anymore?

File: src/__internal__/focus-trap/focus-trap.component.js [hunk 9]
```diff
@@ -13,19 +14,21 @@ import {
...
 } from "./focus-trap-utils";
+import { ModalContext } from "../../components/modal/modal.component";
 
 const FocusTrap = ({
   children,
   autoFocus = true,
   focusFirstElement,
   bespokeTrap,
   wrapperRef,
+  isAnimationComplete = false,
```
Comment (edleeks87) [inline src/__internal__/focus-trap/focus-trap.component.js:25]: Is this used?

File: src/__internal__/focus-trap/focus-trap.component.js [hunk 10]
```diff
@@ -54,11 +57,22 @@ const FocusTrap = ({
...
+    ) {
       setElementFocus(focusFirstElement || firstElement);
       firstOpen.current = false;
     }
-  }, [autoFocus, firstElement, focusFirstElement]);
+  }, [
+    autoFocus,
+    firstElement,
+    focusFirstElement,
+    isAnimationComplete,
```
Comment (edleeks87) [inline src/__internal__/focus-trap/focus-trap.component.js:73]: Is this prop used at all?

File: src/__internal__/focus-trap/focus-trap.component.js [hunk 11]
```diff
@@ -13,19 +14,21 @@ import {
...
   bespokeTrap,
   wrapperRef,
+  isAnimationComplete = false,
 }) => {
   const firstOpen = useRef(true);
   const [focusableElements, setFocusableElements] = useState();
   const [firstElement, setFirstElement] = useState();
   const [lastElement, setLastElement] = useState();
-
+  const modalContext = useContext(ModalContext);
```
Comment (edleeks87) [inline src/__internal__/focus-trap/focus-trap.component.js:31]: You could destructure the value from the context if you wanted

```suggestion
  const { isAnimationComplete } = useContext(ModalContext);
```

File: src/components/modal/modal.component.js [hunk 12]
```diff
@@ -7,6 +7,8 @@ import StyledPortal from "../portal/portal.style";
 import ModalManager from "./__internal__/modal-manager";
 import { StyledModal, StyledModalBackground } from "./modal.style";
 
+export const ModalContext = React.createContext();
```
Comment (edleeks87) [inline src/components/modal/modal.component.js:10]:
```suggestion
export const ModalContext = React.createContext({});
```

File: src/__internal__/focus-trap/focus-trap.component.js (hunk 9 above)
Comment (sjohnson-sage) [reply src/__internal__/focus-trap/focus-trap.component.js]: How can it be legacy when you've added it here?

**Comment by dawidzarzycki** [comment]
> Please link this to the issues so they get closed automatically.

good point! Fixed

**Review by edleeks87 — APPROVED** [review]

**Review by edleeks87 — APPROVED** [review]

**Review by sjohnson-sage — APPROVED** [review]

**Review by ianforster-sage — APPROVED** [review]

**Review by sianford — APPROVED** [review]

**Comment by sianford** [comment]
Tested in Safari, Chrome, Firefox and Edge. Testing passed ✅

**Pull request merged**